Use swagger-ui-dist instead of swagger-ui

thesocialdev opened the pull request:
Based on hyperswitch [implementation](https://github.com/wikimedia/hyperswitch/commit/a040a59f2f1f07f95471235a9a5cbbdf9ab42e42
).

Bug: [T218217](https://phabricator.wikimedia.org/T218217)

thesocialdev committed:
Use swagger-ui-dist instead of swagger-ui

Based on hyperswitch implementation. See
https://github.com/wikimedia/hyperswitch/commit/a040a59f2f1f07f95471235a9a5cbbdf9ab42e42

Bug: T218217

Pchelolo commented:
1. Let's bump minor version
2. For this PR, let's also remove `node` from .travis.yaml as a version, node 12 just released and nothing works with it, fixing node 12 support will be a separate project.

thesocialdev committed:
Hygiene and remove swagger 2 support

1) Removing 'node' from travis.yml, support node 10 instead of node 8
2) Support only openapi 3 spec
3) Update spec.example.yaml to be openapi 3 compliant
4) Bump minor version

d00rman commented:
> 1. Let's bump minor version

Actually, let's not do that just yet, let's get rid of the optional params first too.

> 2. For this PR, let's also remove `node` from .travis.yaml as a version, node 12 just released and nothing works with it, fixing node 12 support will be a separate project.

Pchelolo commented:
> Actually, let's not do that just yet, let's get rid of the optional params first too.

We should do it as a part of the same PR actually..

d00rman commented:
> We should do it as a part of the same PR actually..

Or that, yes, even better. @mateusbs17 could you adjust the routes in [`spec.template.yaml`](https://github.com/wikimedia/service-template-node/blob/2a449f7e27ed2e23e182d3ecbf11fd300e9390e1/spec.template.yaml#L90) and split it in two (one with and the other without the extra param), please? Also, please do it in the same commit that bumps the version.

thesocialdev commented:
> > We should do it as a part of the same PR actually..
> 
> Or that, yes, even better. @mateusbs17 could you adjust the routes in [`spec.template.yaml`](https://github.com/wikimedia/service-template-node/blob/2a449f7e27ed2e23e182d3ecbf11fd300e9390e1/spec.template.yaml#L90) and split it in two (one with and the other without the extra param), please? Also, please do it in the same commit that bumps the version.

Ok. 

One question, should we remove specific code that checks for swagger 2.0 spec versions and keep only openapi 3?

Pchelolo commented:
Hm I see in MCS patch you are switching to openApi 3 right away. I think given that this is a template, let's be bold and completely switch to openapi 3 here too? What do you think @d00rman ?

d00rman commented:
> let's be bold and completely switch to openapi 3 here too?

+1, let's do it. Apart from the optional path param, that we need to get rid of anyway, there's nothing that ties us to v2 in the spec template.

d00rman commented on .travis.yml:
```diff
@@ -5,4 +5,3 @@ sudo: false
 node_js:
   - "6"
   - "8"
-  - "node"
```
Since we are changing stuff here, let's remove node 8 and have node 10.

d00rman commented on app.js:
```diff
@@ -80,7 +80,8 @@ function initApp(options) {
     }
-    if (!app.conf.spec.swagger) {
+    if (!app.conf.spec.swagger && !app.conf.spec.openapi) {
+        // Set default to swagger 2.0 for backwards compability
```
Let's set the def to openapi 3

berndsi commented on lib/swagger-ui.js:
```diff
@@ -27,25 +27,48 @@ function processRequest(app, req, res) {
+                .scheme-container {
+                    display: none
+                }
+            `;
```
Some of this functionality seems like it could be a good candidate for a common npm library instead of duplicating the same thing throughout our various Node services.

Pchelolo replied:
you are reading my mind @berndsi 

@mateusbs17 would you mind filing a ticket to abstract this out?

thesocialdev replied:
Maybe the whole lib folder inside service-template-node would be a good fit for common libs, I filled the ticket at https://phabricator.wikimedia.org/T222457

Pchelolo commented on test/features/app/spec.js:
```diff
@@ -256,7 +256,7 @@ describe('Swagger spec', function() {
         }
         // check the high-level attributes
-        ['info', 'swagger', 'paths'].forEach((prop) => {
+        ['info', 'openapi', 'paths'].forEach((prop) => {
```
Let's use openapi-schema-validator in this test just like RESTBase does.

Pchelolo commented:
I'm gonna merge this and do a couple of followups
